sputnikdao2/src/proposals.rs

```diff
@@ -217,41 +218,218 @@ impl From<ProposalInput> for Proposal {
     }
 }
 
+// TODO: Use near_contract_standards::storage_management::StorageBalance.
+// The original 'StorageBalance' struct does not support deserialization. Delete this
+// and use the original one after https://github.com/near/near-sdk-rs/pull/630 gets released.
+#[derive(BorshDeserialize, BorshSerialize, Serialize, Deserialize)]
+#[serde(crate = "near_sdk::serde")]
+pub struct StorageBalance {
+    pub total: U128,
+    pub available: U128,
+}
+
+#[ext_contract(ext_storage_management)]
+pub trait StorageManagement {
+    fn storage_deposit(
+        &mut self,
+        account_id: Option<AccountId>,
+        registration_only: Option<bool>,
+    ) -> StorageBalance;
+    fn storage_balance_of(&self, account_id: AccountId) -> Option<StorageBalance>;
+}
+
+#[ext_contract(ext_self)]
+pub trait ExtSelf {
+    fn callback_after_storage_deposit(
+        &mut self,
+        token_id: AccountId,
+        proposer_account: AccountId,
+        receiver_id: AccountId,
+        amount: U128,
+        attached_deposit: U128,
+        memo: String,
+        msg: Option<String>,
+    ) -> PromiseOrValue<()>;
+    fn callback_after_storage_balance_of(
+        &mut self,
+        token_id: AccountId,
+        proposer_account: AccountId,
+        receiver_id: AccountId,
+        amount: U128,
+        attached_deposit: U128,
+        memo: String,
+        msg: Option<String>,
+    ) -> PromiseOrValue<()>;
+}
+
+#[near_bindgen]
+impl Contract {
+    #[allow(dead_code)]
+    #[private]
+    pub fn callback_after_storage_deposit(
+        &mut self,
+        token_id: AccountId,
+        proposer_account: AccountId,
+        receiver_id: AccountId,
+        amount: U128,
+        attached_deposit: U128,
+        memo: String,
+        msg: Option<String>,
+    ) -> PromiseOrValue<()> {
+        assert_eq!(
+            env::promise_results_count(),
+            1,
+            "ERR_UNEXPECTED_CALLBACK_PROMISES"
+        );
+        match env::promise_result(0) {
+            PromiseResult::NotReady => {
+                Promise::new(proposer_account.clone()).transfer(attached_deposit.0);
+                panic!("storage_deposit Error: Received PromiseResult::NotReady")
+            }
+            PromiseResult::Failed => {
+                Promise::new(proposer_account.clone()).transfer(attached_deposit.0);
+                panic!("storage_deposit Error: Received PromiseResult::Failed")
```

Comment:
For this case and the `PromiseResult::NotReady` above, from some tests I made it appears that Promises that are being created in a function are discarded if a panic happens. So I think the panics should be avoided in there, and if a failure signal is necessary, perhaps a bool can indicate it?